### ISSUE TITLE
Rename add(observer:) function on Procedure in regard of Swift 3

### ISSUE THREAD

**Issue by martnst** [ProcedureKit/ProcedureKit#507]
In my understanding the add functions don't need a named parameter in Swift 3 unless it is adding some notion of understanding what the method does. 

# Dependency 
`func add<Dependency: ProcedureProtocol>(dependency: Dependency)` 

I think it makes sense to specify that another operation is added as dependency. Just adding a operation to another would not make sense without this context. 

**Example:** 
```swift
myOperation.add(dependency: yourOperation)  // perfectly fine 
```

# Observer 
`func add<Observer: ProcedureObserver>(observer: Observer) where Observer.Procedure == Self` 

I think it's not adding any important details for the adding of observers though. 

**Example:** 
```swift
myOperation.add(observer: MyObserver())    // duplicated bbserver
myOperation.add(MyObserver())              // reads better 
```

**Comment by danthorpe** [comment]
Hi, sorry @martnst - I don't agree. If you are instantiating the observer in place, then more than likely the name of the observer type will disambiguate for the reader. However, not so in situations where a previously defined observer value is added.

From the Swift 3 naming guidelines [fundamentals](https://swift.org/documentation/api-design-guidelines/#fundamentals):

> Fundamentals
>
> Clarity at the point of use is your most important goal. Entities such as methods and properties are declared only once but used repeatedly. Design APIs to make those uses clear and concise. When evaluating a design, reading a declaration is seldom sufficient; always examine a use case to make sure it looks clear in context.
> 
> Clarity is more important than brevity. Although Swift code can be compact, it is a non-goal to enable the smallest possible code with the fewest characters. Brevity in Swift code, where it occurs, is a side-effect of the strong type system and features that naturally reduce boilerplate.
>

**Comment by martnst** [comment]
Okay, fair enough. However, in the WWDC talk about Swift 3 they argued that you should have a clear variable name which would indicated that it's referring to an observer.

**Comment by danthorpe** [comment]
Well, I think the Swift team is fantastic, and they have some great ideas. But the reality is that most developers have not seen this session, and of those that have, probably don't recall much more than a screen of bananas 😄

The first two _fundamental_ rules pretty much say that writing compact code for the sake of it, is a non-goal.

Given that there will likely be three APIs called `add-something` - I don't see why they would have a different format, and I don't think they should be named the same either.

**Comment by jshier** [comment]
True, but given that the guidelines provide no actual guidance for when functions of the same name perform different actions, we're left to improvise. There seem to be 3 options here:

1. `add(_ x:)`
2. `add(x:)`
3. `addX(_ x:)`

Now, if we don't need disambiguation at the call sites, then perhaps 1 is the best option. However, besides disambiguation we also need to consider the principle of least surprise. Would adding a condition, observer, or dependency using functions of the same name be surprising? I'd say probably, at least for developers unfamiliar with the framework. Given that, having a label is probably the better option.

Additionally, it may be important, through the function name, to indicate that some of these must be added before the procedure starts, which observers can be added while it's running. I'm not sure what that would look like.

**Comment by danthorpe** [comment]
@jshier your last point here, was actually sort of what led me to changing `add(condition: C)` to `attach(condition: C)` - as it's very important to add conditions to procedures before the procedure is added to the queue. Whereas for observers, as you say, not necessary.

And in terms of english grammar (again, argued as the _thing to do_ in the Swift 3 WWDC session), we "attach conditions to things". But, I'm not sure about it, and the API for dependencies is `add(dependency:)` which must also be done before the receiver is added to the queue.

So... I'm leaning towards having all three be "add", with labels. But, definitely open to suggestions on renaming this if there is a more appropriate verb. 

Thanks for kicking this off @martnst!

**Comment by martnst** [comment]
I don't think it's possible to over the timing / order of calling thing in a method name. I great documentation and error handling is the only way cover this properly.

**Comment by danthorpe** [comment]
Yeah, but the name of an API _is_ the first element of documentation 😉

**Comment by danthorpe** [comment]
Gonna close this now - not renaming `add(observer:` but have renamed to `add(condition:` to be matching.
